icu4c/source/common/locbased.cpp

```diff
@@ -27,29 +28,38 @@ const char* LocaleBased::getLocaleID(ULocDataLocaleType type, UErrorCode& status
 
     switch(type) {
     case ULOC_VALID_LOCALE:
-        return valid;
+        return valid->data();
     case ULOC_ACTUAL_LOCALE:
-        return actual;
+        return actual->data();
     default:
         status = U_ILLEGAL_ARGUMENT_ERROR;
         return nullptr;
     }
 }
 
-void LocaleBased::setLocaleIDs(const char* validID, const char* actualID) {
+void LocaleBased::setLocaleIDs(const char* validID, const char* actualID, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
     if (validID != nullptr) {
-      uprv_strncpy(valid, validID, ULOC_FULLNAME_CAPACITY);
-      valid[ULOC_FULLNAME_CAPACITY-1] = 0; // always terminate
+      valid->clear();
+      valid->append(validID, -1, status);
```

Comment:
Hm. We should really have a CharString function overload
```
    CharString &copyFrom(StringPiece s, UErrorCode &errorCode);
```
to simplify such call sites. In particular, that would eliminate error checking because it wouldn't clear & replace when there is already a failure.

I can work on that as a follow-up if you like.